From a98537ae03b2f94310939df84d5202a178538354 Mon Sep 17 00:00:00 2001
From: clodeindustrie <clodeindustrie@gmail.com>
Date: Fri, 13 Oct 2023 12:16:29 +1100
Subject: [PATCH] Add ns-set-application-icon

---
 src/nsfns.m | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/src/nsfns.m b/src/nsfns.m
index 038a3fa23ad..b38796f0592 100644
--- a/src/nsfns.m
+++ b/src/nsfns.m
@@ -2141,7 +2141,30 @@ Frames are listed from topmost (first) to bottommost (last).  */)
   return Qnil;
 }
 
+DEFUN ("ns-set-application-icon", Fns_set_application_icon, Sns_set_application_icon, 1, 1, 0,
+       doc: /* Set the running application's icon to PATH,
+               and return true, or return nil if it fails.  */)
+     (Lisp_Object path)
+{
+  CHECK_STRING (path);
+  char *s = SSDATA(path);
+  NSString *ns = [[NSString stringWithUTF8String:s] stringByExpandingTildeInPath];
+  NSURL *url = [NSURL fileURLWithPath:ns];
+  if(!url) {
+    NSLog(@"could not create url for path: %@", ns);
+    return Qnil;
+  }
+  NSImage *img = [[NSImage alloc] initWithContentsOfURL: url];
+  if(!img) {
+    NSLog(@"could not create image from url: %@", url);
+    return Qnil;
+  }
+  NSLog(@"setting application icon");
+  [NSApp setApplicationIconImage:img];
+  return Qt;
+}
+
+  defsubr (&Sns_set_application_icon);
 
 DEFUN ("x-display-list", Fx_display_list, Sx_display_list, 0, 0, 0,
        doc: /* SKIP: real doc in xfns.c.  */)
--
2.32.1 (Apple Git-133)

